Feature/code clean up

dav009 opened the pull request:
So this is a big code clean up of the model editor:
- Styled the code a bit using:
  - [scalariform](https://github.com/mdr/scalariform) and [scalastyle](http://www.scalastyle.org/command-line.html)
  - as well as following: http://docs.scala-lang.org/style/
  - There are few warnings  still thrown by scalastyle (i.e: using `return` statements) I think we can remove them as we go on refactoring other bits

----UPDATE----

The other relevant changes are:
- Stores are now loaded in a lazy way. This should improve speed when only working with a particular store
- Only stores which are used are exported. This should also improve speed
- Renamed the store classes
- Replaced some of the `for loops` with `foreach` loops
- Using `scopt` library to create command-line parser
- Command-line interface has changed( check readme for details) 
  - command : `surfaceform`, with the following actions:
    - `make-spottable`
    - `make-no-spottable`
    - `candidates`
    - `stats`
      - `copy-candidates`
  - command: `topic`, with the following actions:
    - `search`
    - `check-context`
    - `clean-set-context`
  - command: `association` with the follownig actions:
    - `remove`
  - command: `file-update` with the following actions:
    - `all`
    - `only-context`
- Refactor the code parsing the input files
- Put all the stores in a separate package

--- Update 2 ---
- Added a maven plugin which generates a script calling the jar. So we avoid writting the same over an over. the script is generated in `target/bin/model-editor`. So now running hte model editor is just: `target/bin/model-editor surfaceform make-spottable pathToModel/model Barack`

I would like to remove the `Updating a model From File (Two Steps)` part in the docs. Since anyway this is meant to run in a big machine, any arguments against it?

---
- [x] far
- [x] Thiango

dav009 committed:
moving classes

dav009 committed:
changing classes packages

dav009 committed:
Big Code Clean

dav009 committed:
more cleaning

dav009 committed:
styling code

dav009 committed:
updating docs

tgalery commented on src/main/scala/org/idio/dbpedia/spotlight/IdioSpotlightModel.scala:
```diff
@@ -3,116 +3,114 @@ package org.idio.dbpedia.spotlight
 /**
  * Created by dav009 on 23/12/2013.
  */
-import org.dbpedia.spotlight.model.{TokenType, OntologyType}
```
I think it's a bit weird the prefix the classes with Idio, since they don't relate to the OO logic. The name of the company is already in the package, so I think it would be better to name this CustomSpotlightModel or something like that. But that's my humble opinion, only.

tgalery commented on src/main/scala/org/idio/dbpedia/spotlight/IdioSpotlightModel.scala:
```diff
@@ -3,116 +3,114 @@ package org.idio.dbpedia.spotlight
-  try{
-    this.idioContextStore =  new IdioContextStore(pathToFolder, idioTokenTypeStore.tokenStore)
+  try {
+    this.idioContextStore = new IdioContextStore(pathToFolder, idioTokenTypeStore.tokenStore)
```
perhaps customContextStore and so on. I'll stop nagging you about this, but bare in mind that if you change this, the changes would have do be consistent throughout.

dav009 replied:
Sure, I agree on your previous point about the names.

tgalery commented on src/main/scala/org/idio/dbpedia/spotlight/IdioSpotlightModel.scala:
```diff
@@ -169,74 +167,50 @@ class IdioSpotlightModel(val pathToFolder:String){
-      case ex: Exception =>{
+    } catch {
+      case ex: Exception => {
         println("\tNot Context Store found....")
```
Perhaps compress these messages in one, and actually print the exception message to see what went wrong exactly.

dav009 replied:
I can print the exception message but I would like to keep all the try/catches there.
The reason for this is that for [dev-testing,  locally seeing if something fixes a model] you can turn "off" certain stores.

tgalery commented:
I mentioned to you in passing that it would be nice to have a too similar to knife. Like the first argument of the editor would an entity (surface form, candidate, or resource) and the next argument would be an action (make-spotable, associate, change context), but i think this would take a while to fix, so maybe in another life we could do this.

dav009 commented:
It doesn't sound so complicated. I will take a look for a  lib similar to python's `optpars`

tgalery commented:
Cool, but it's still a minor point, so up to you whether you fancy doing that.

dav009 committed:
renaming classes and variables

dav009 committed:
renaming classes/variables

dav009 committed:
improving command line tool

dav009 committed:
allowing files as input

dav009 committed:
lazy loading/exporting stores

dav009 committed:
updating docs

dav009 committed:
updating association command

dav009 committed:
fixing a few bugs

dav009 commented:
@tgalery  you can rereview!

dav009 committed:
little doc fix

tgalery commented on pom.xml:
```diff
@@ -52,10 +52,17 @@
 
 
-    <!-- Test -->
+      <!-- Test -->
```
why the extra space?

dav009 replied:
cause I absolutely forgot these poms

tgalery commented on src/main/scala/org/idio/dbpedia/spotlight/CustomSpotlightModel.scala:
```diff
@@ -0,0 +1,449 @@
+        this.customCandidateMapStore.createCandidateMapForSurfaceForm(surfaceFormId, new Array[Int](0), new Array[Int](0)))
+  }
+  /*
+    takes all topic candidates for the surfaceForm1
```
takes => Takes

tgalery commented on src/main/scala/org/idio/dbpedia/spotlight/CustomSpotlightModel.scala:
```diff
@@ -0,0 +1,449 @@
+    writer.close()
+  }
+
+}
```
EOF

tgalery commented on src/main/scala/org/idio/dbpedia/spotlight/stores/CustomCandidateMapStore.scala:
```diff
@@ -0,0 +1,180 @@
+package org.idio.dbpedia.spotlight.stores
+
+/**
+ * Created by dav009 on 23/12/2013.
```
Perhaps add a more robust information in the docs. Like @author and your email ?

tgalery commented on src/main/scala/org/idio/dbpedia/spotlight/stores/CustomCandidateMapStore.scala:
```diff
@@ -0,0 +1,180 @@
+  }
+
+  /*
+  * increments the candidates Counts for a given surfaceForm and candidate
```
captalize 'i'

tgalery commented on src/main/scala/org/idio/dbpedia/spotlight/stores/CustomDbpediaResourceStore.scala:
```diff
@@ -0,0 +1,90 @@
+
+class CustomDbpediaResourceStore(val pathtoFolder: String) {
+
+  var resStore: MemoryResourceStore = MemoryStore.loadResourceStore(new FileInputStream(new File(pathtoFolder, "res.mem")))
```
does this conform to the guidelines ? Maybe split this statement in multiple lines?

dav009 committed:
script calling the edito

tgalery commented on src/main/scala/org/idio/dbpedia/spotlight/stores/CustomDbpediaResourceStore.scala:
```diff
@@ -0,0 +1,90 @@
+  * i.e: [dbpdia:person, dbpedia:location] => [100, 392]..
+  * */
+  def getTypesIds(dbpediaTypes: Array[String]): Array[Array[java.lang.Short]] = {
+    var dbpediaTypesForResource2 = new Array[Array[java.lang.Short]](1)
```
would this work for non quantized versions of the model ?

dav009 replied:
so, yes, the type ids are by default shorts, so to what this line is concerned it works in both  0.6 and 0.7 versions. but this whole branch only works for 0.6 models

tgalery commented on src/main/scala/org/idio/dbpedia/spotlight/utils/ModelFileParser.scala:
```diff
@@ -0,0 +1,144 @@
+    return (this.setOfUCSurfaceForms, setOfLCSurfaceForms, this.setOfDbpediaURIS, lowerCasesMap, this.parsedLines, this.setOfContextWords)
+  }
+
+}
```
EOF

tgalery commented on src/main/scala/org/idio/dbpedia/spotlight/utils/ModelFileParser.scala:
```diff
@@ -0,0 +1,144 @@
+
+    source.close()
+
+    return (this.setOfUCSurfaceForms, setOfLCSurfaceForms, this.setOfDbpediaURIS, lowerCasesMap, this.parsedLines, this.setOfContextWords)
```
Line too long?

tgalery commented on src/main/scala/org/idio/dbpedia/spotlight/utils/ModelUpdateFromFile.scala:
```diff
@@ -1,122 +1,66 @@
+    println("Parsing INPUT-FILE")
 
-    var idioSpotlightModel:IdioSpotlightModel = new IdioSpotlightModel(this.pathToModelFolder)
+    val (setOfUpperCaseSF, setOfLowerCaseSF, setOfDbpediaURIS, lowerSfMap, parsedLines, setOfContextWords) = fileParser.parseFile()
```
Line too long

dav009 committed:
typos/long lines fixes

dav009 committed:
adding header license

dav009 committed:
styling

dav009 committed:
adding notice and license files

dav009 commented:
- Fixed line lengths
- Added Headers with candidate license
- Added Notice/License Files
- Specified licenses  of dependencies in POM file

@tgalery ready for re-re-re-re-re-re-re....-re review

dav009 committed:
dependencies licenses in POM

dav009 committed:
adding licenses to NOTICE

tgalery commented:
Looks good :+1:  :sake: . The only tiny thing, should we have a license.txt file linked from within the readme. Also, a few suggestions for actions change.

`make-spottable` => `make-spotable` (I think spotable is with only one 't' but i m not sure)
`make-no-spottable` => `make-unspotable`
`only-context` => `context-only`

Is that ok ?

dav009 commented:
sure.
`spottable` [seems to be alright](http://www.thefreedictionary.com/spottable)

There is a link in readme.md to the standard `apache 2.0` license, which is what I usually saw in the readme's of the big companies.

dav009 committed:
updating command names

dav009 committed:
renamed make-unspottable command

papaonlegs commented on NOTICE:
```diff
@@ -0,0 +1,28 @@
+https://code.google.com/p/mockito/
+
+JUnit
+License: Eclipse Public License - v 1.0
```
You can just put `Eclipse Public License - v 1.0`

papaonlegs commented on NOTICE:
```diff
@@ -0,0 +1,28 @@
+https://github.com/junit-team/junit
+
+Scala-Test
+Apache Software License, Version 2.0
```
Is there a difference between `Apache Software License, Version 2.0` and `Apache Public License 2.0`? I couldn't find any

dav009 replied:
:sake:

papaonlegs commented on README.md:
```diff
@@ -1,13 +1,20 @@
+  - Create associations between surface forms and dbpedia uris
+  - Remove associations between surface forms and dbpedia uris
+  - Make surface forms spottable
+  - Make surface forms no spottable
```
`unspottable`

papaonlegs commented on README.md:
```diff
@@ -1,13 +1,20 @@
 - (Oracle) Java 1.7
-- Scala >= 2.9.x
-- Compiling DBPedia Spotlight from source (see below)
+- Scala 2.9.x (If you want to run it interatively from a terminal)
```
`interactively`

papaonlegs commented on README.md:
````diff
@@ -47,13 +54,16 @@ Turkish (located in the `tr` folder). Feel free to play with other languages, if
 
 ```
-java -Xmx4000M -jar  target/idio-spotlight-model-0.1.0-jar-with-dependencies.jar explore path-to-turkish/tr/model/
+sh target/bin/model-editor explore path-to-turkish/tr/model/
````
turkish?

dav009 replied:
it is just an example, but yes there is a turkish model, very small though.

papaonlegs commented on README.md:
````diff
@@ -78,82 +95,171 @@ java -Xmx15360M -Xms15360M -jar idio-spotlight-model-0.1.0-jar-with-dependencies
 example:
 ```
-java -Xmx4000M -jar  target/idio-spotlight-model-0.1.0-jar-with-dependencies.jar explore path-to-turkish/tr/model/
+sh target/bin/model-editor explore path-to-turkish/tr/model/
````
Turkish model again?

papaonlegs commented on README.md:
```diff
@@ -78,82 +95,171 @@ java -Xmx15360M -Xms15360M -jar idio-spotlight-model-0.1.0-jar-with-dependencies
+- **subcommand**: `clean-set-context`
+- **arg1**: `pathToSpotlightModel/model`
+- **arg2**: pathToFile
+- **result**: The context words and counts for the topics in the file will be cleared. The specified context Words will be stemmed and added with their respect counts to the context vector of the given topics.
```
respective

papaonlegs commented on README.md:
````diff
@@ -78,82 +95,171 @@ java -Xmx15360M -Xms15360M -jar idio-spotlight-model-0.1.0-jar-with-dependencies
+each line of the given input file should be like: 
+
+```
+dbpediaUri <tab> contextWordsSeparatedByPipe <tab> countsSeparatedBytab
````
Might put a note that counts and context words should be same number, if they should?

dav009 replied:
yeah, and actually there is a clear error here `countsSeparatedBytab` should be `countsSeparatedByPipe`

papaonlegs commented on README.md:
```diff
@@ -78,82 +95,171 @@ java -Xmx15360M -Xms15360M -jar idio-spotlight-model-0.1.0-jar-with-dependencies
+ - `stats` : printing stats of a surface form
+ - `candidates` : printing the list of candidates of a surface form
+ - `make-spottable` : making surfaceforms spottable
+ - `make-unspottable` : making surfaceforms no spottable
```
`unspottable`

papaonlegs commented on README.md:
```diff
@@ -78,82 +95,171 @@ java -Xmx15360M -Xms15360M -jar idio-spotlight-model-0.1.0-jar-with-dependencies
+would check the candidate topics for the surface form `evrimleri`
 
-example: 
+### Making a list of Surface Forms not Spottable
```
`Unspottable`

papaonlegs commented on README.md:
```diff
@@ -219,18 +325,20 @@ steps 5-6 could be applied while ignoring 1-4 when:
 
 Once you start a scala console you can use it like `ipython` to create instances of the scala classes we have, to load the models, check if dbpedia id's exist, add new dbpedia ids, add new surface forms etc..
 
+do:  `:cp pathTo/ModelEditor.jar`
```
Will this immediately eat up ~15 gig and how long does it take? Maybe put a notice at the top of this file that the script will eat up X memory when running if it does, or something.

papaonlegs commented on src/main/scala/org/idio/dbpedia/spotlight/CustomSpotlightModel.scala:
```diff
@@ -0,0 +1,466 @@
+    val surfaceFormID: Int = this.customSurfaceFormStore.getAddSurfaceForm(surfaceFormText)
+    this.customSurfaceFormStore.boostCountsIfNeeded(surfaceFormID)
+    var defaultSupportForDbpediaResource: Int = 11
+    val defaultSupportForCandidate: Int = 30
```
Comment on why these values for default?

dav009 replied:
when you call the spotlight endpoint you can set certain filters , for example filtering topics below a certain support value.
These values are above the default values for those filters.

papaonlegs replied:
A comment like that in the code would be great, so people that don't know about the spotlight filter values.

papaonlegs commented on src/main/scala/org/idio/dbpedia/spotlight/CustomSpotlightModel.scala:
```diff
@@ -0,0 +1,466 @@
+        val dbpediaResource = this.customDbpediaResourceStore.resStore.getResource(candidate)
+        avgSupport = avgSupport + dbpediaResource.support
+      }
+      val calculatedSupport = (avgSupport / candidates.length).toInt
```
I think naming-wise `avgSupport` could be `sumSupport` and `calculatedSupport` could be `avgSupport` unless you want to `avgSupport = (avgSupport / candidates.length).toInt`. variable reuse ++

papaonlegs commented on src/main/scala/org/idio/dbpedia/spotlight/CustomSpotlightModel.scala:
```diff
@@ -0,0 +1,466 @@
+        defaultSupportForDbpediaResource = calculatedSupport
+      }
+
+      avgSupportCandidate = math.max(defaultSupportForCandidate, this.customCandidateMapStore.getAVGSupportForSF(surfaceFormID)) + 10
```
Comment on why + 10?

papaonlegs commented on src/main/scala/org/idio/dbpedia/spotlight/CustomSpotlightModel.scala:
```diff
@@ -0,0 +1,466 @@
+    }
+
+    // create or get the dbpedia Resource
+    val dbpediaResourceID: Int = this.customDbpediaResourceStore.getAddDbpediaResource(candidateURI, defaultSupportForDbpediaResource, types)
```
Does types have to be ordered in a specific way?

dav009 replied:
no that I remember, however we are ignoring types at the moment

papaonlegs replied:
Oh, Cool.

papaonlegs commented on src/main/scala/org/idio/dbpedia/spotlight/CustomSpotlightModel.scala:
```diff
@@ -0,0 +1,466 @@
+
+  /*
+  * Removes all the context words and context counts of a dbepdia topic
+  * and sets the context words and cotnext counts specified in the command line.
```
`context`

papaonlegs commented on src/main/scala/org/idio/dbpedia/spotlight/CustomSpotlightModel.scala:
```diff
@@ -0,0 +1,466 @@
+  def getStatsForSurfaceForm(surfaceFormText: String) {
+    val surfaceForm = this.customSurfaceFormStore.sfStore.getSurfaceForm(surfaceFormText)
+
+    println("surface form id:" + surfaceForm.id)
```
Capitalize?

papaonlegs replied:
And the other strings here too?

papaonlegs commented on src/main/scala/org/idio/dbpedia/spotlight/CustomSpotlightModel.scala:
```diff
@@ -0,0 +1,466 @@
+  }
+
+  /*
+  * Prints the first 40 surface forms and their respective candidates
```
Maybe allow an input limit with default to 40 if not provided?

dav009 commented on src/main/scala/org/idio/dbpedia/spotlight/CustomSpotlightModel.scala:
```diff
@@ -0,0 +1,466 @@
+    var counts: Array[Int] = this.customContextStore.contextStore.counts(dbpediaResourceID)
+    println("Contexts for " + dbpediaResourceURI + " Id:" + dbpediaResourceID)
+    for (i <- 0 to tokens.size - 1) {
+      this.customTokenTypeStore.tokenStore.idFromToken
```
remove this

papaonlegs commented on src/main/scala/org/idio/dbpedia/spotlight/CustomSpotlightModel.scala:
```diff
@@ -0,0 +1,466 @@
+  }
+
+  /*
+  *  Makes a SF not spottable by reducing its annotationProbability to 0.1
```
`unspottable`

papaonlegs commented on src/main/scala/org/idio/dbpedia/spotlight/CustomSpotlightModel.scala:
```diff
@@ -0,0 +1,466 @@
+  }
+
+  /*
+*  Makes a SF  spottable by reducing its annotationProbability to 0.1
```
There's two spaces between SF and spottable

papaonlegs commented on src/main/scala/org/idio/dbpedia/spotlight/CustomSpotlightModel.scala:
```diff
@@ -0,0 +1,466 @@
+  def copyCandidates(surfaceTextSource: String, surfaceTextDestination: String) {
+
+    val sourceSurfaceForm = this.customSurfaceFormStore.sfStore.getSurfaceForm(surfaceTextSource)
+    val destinySurfaceForm = this.customSurfaceFormStore.sfStore.getSurfaceForm(surfaceTextDestination)
```
`destinationSurfaceForm` ?

papaonlegs commented on src/main/scala/org/idio/dbpedia/spotlight/CustomSpotlightModel.scala:
```diff
@@ -0,0 +1,466 @@
+  }
+
+  /*
+  * Removes the link between a SF and a Dbpedia Topic
```
an\* SF

papaonlegs commented on src/main/scala/org/idio/dbpedia/spotlight/CustomSpotlightModel.scala:
```diff
@@ -0,0 +1,466 @@
+  }
+
+  /**
+   *  given a SF returns the list of candidate Topics
```
Given\* an\* SF return\* ...

papaonlegs commented on src/main/scala/org/idio/dbpedia/spotlight/CustomSpotlightModel.scala:
```diff
@@ -0,0 +1,466 @@
+
+  /*
+  * Adds a set of tokens to the token type store.
+  * It only generte reverse look ups once.
```
generte? generates?

papaonlegs commented on src/main/scala/org/idio/dbpedia/spotlight/CustomSpotlightModel.scala:
```diff
@@ -0,0 +1,466 @@
+
+      } catch {
+        case e: Exception => {
+          println("\t not found context for" + dbpediaTopicID)
```
`Context not found for`*

papaonlegs commented on src/main/scala/org/idio/dbpedia/spotlight/CustomSpotlightModel.scala:
```diff
@@ -0,0 +1,466 @@
+        for ((tokenType, count) <- contextCounts) {
+          lineInformation += tokenType.tokenType + ":" + count
+        }
+        val writeLine = lineInformation.mkString("\t") + "\n"
```
Just out of curiosity why did you tab delimit here, instead of pipe?

dav009 replied:
no specific reason. This was used only at the beginning when we wanted to know the distribution of context words/ counts, sort of having a human-readable version of the context store.

papaonlegs replied:
Cool

papaonlegs commented on src/main/scala/org/idio/dbpedia/spotlight/SpotlightModelReader.scala:
```diff
@@ -1,180 +1,225 @@
-      action match{
+    parsingResult match {
+      case Some(commandLineConfig) => runCommand(commandLineConfig)
+      case None => println("please enter a valid command...")
```
Please*

papaonlegs commented on src/main/scala/org/idio/dbpedia/spotlight/SpotlightModelReader.scala:
```diff
@@ -1,180 +1,225 @@
+    val subCommand: String = commandLineConfig.commands.lift(1).getOrElse("")
+    val pathToModelFolder: String = commandLineConfig.pathToModelFolder
+
+    lazy val spotlightModelReader = Main.getSpotlightModel(commandLineConfig.pathToModelFolder)
```
May as well use `pathToModelFolder` here, since you just created the val.

papaonlegs commented on src/main/scala/org/idio/dbpedia/spotlight/SpotlightModelReader.scala:
```diff
@@ -1,180 +1,225 @@
+            sourceFile.getLines().toArray.map{ line =>
+              val surfaceForm = line.trim()
+              surfaceForm
+            }
```
Can you do this here?

`sourceFile.getLines().toArray.map{ line =>line.trim()}`

dav009 replied:
yes, even shorter:

```
sourceFile.getLines().toArray.map{ _.trim()}
```

papaonlegs replied:
Like that doesn't need the curly braces, right? Normal brackets should work?

dav009 replied:
yeah `()` would work

papaonlegs commented on src/main/scala/org/idio/dbpedia/spotlight/SpotlightModelReader.scala:
```diff
@@ -1,180 +1,225 @@
+        surfaceTexts.foreach{ surfaceForm:String =>
+          println("making  spottable:"+ surfaceForm)
         }
+        surfaceTexts.foreach(spotlightModelReader.makeSFSpottable)
```
This and the loop above can be one?

```
surfaceTexts.foreach{ surfaceForm:String =>
          println("making  spottable:"+ surfaceForm)
          spotlightModelReader.makeSFSpottable
          }
```

papaonlegs commented on src/main/scala/org/idio/dbpedia/spotlight/SpotlightModelReader.scala:
```diff
@@ -1,180 +1,225 @@
+           }else{
+             val sourceFile = scala.io.Source.fromFile(commandLineConfig.argument)
+             sourceFile.getLines().toArray.map{ line =>
+               val surfaceForm = line.trim()
```
Same comment as above

papaonlegs commented on src/main/scala/org/idio/dbpedia/spotlight/SpotlightModelReader.scala:
```diff
@@ -1,180 +1,225 @@
-          }
+      // makes a piped(|) separated list of SF not spottable.
+      // this is done reducing its annotationProbability
+      case ("surfaceform", "make-unspottable") => {
```
I'm not sure if this is possible but is there any way this case and that for `make-spottable` can be merged, since they're doing the same thing?
e.g

`case ("surfaceform", "make-unspottable" | "make-spottable") => {`

May not be possible, though :(

papaonlegs commented on src/main/scala/org/idio/dbpedia/spotlight/SpotlightModelReader.scala:
```diff
@@ -1,180 +1,225 @@
+        surfaceTexts.foreach{ surfaceForm:String =>
+          println("making not spottable:"+ surfaceForm)
         }
+        surfaceTexts.foreach(spotlightModelReader.makeSFNotSpottable)
```
Same as above, two loops might be mergeable, but maybe you'll just merge the cases

papaonlegs commented on src/main/scala/org/idio/dbpedia/spotlight/SpotlightModelReader.scala:
```diff
@@ -1,180 +1,225 @@
-          spotlightModelReader.replaceAllContext(dbpediaURI, contextWords, contextCounts)
-          println("exporting new model.....")
-          spotlightModelReader.exportModels(pathToModelFolder)
+      // copies the candidates of a sf to another sf
```
an\* sf

papaonlegs commented on src/main/scala/org/idio/dbpedia/spotlight/SpotlightModelReader.scala:
```diff
@@ -1,180 +1,225 @@
+        val sourceFile = scala.io.Source.fromFile(commandLineConfig.argument)
+        sourceFile.getLines().foreach{ line =>
+          val Array(sourceSurfaceForm, destinySurfaceForm) = line.trim().split("\t")
+          spotlightModelReader.copyCandidates(sourceSurfaceForm, destinySurfaceForm)
```
destinationSurfaceForm?

papaonlegs commented on src/main/scala/org/idio/dbpedia/spotlight/SpotlightModelReader.scala:
```diff
@@ -1,180 +1,225 @@
-          }
+      /*
+      * Removes all the context words and context counts of a dbepdia topic
+      * and sets the context words and cotnext counts specified in the command line
```
context*

papaonlegs commented on src/main/scala/org/idio/dbpedia/spotlight/SpotlightModelReader.scala:
```diff
@@ -1,180 +1,225 @@
+        sourceFile.getLines().toArray.foreach{ line =>
+          val Array(dbpediaURI, contextWords, contextCounts) = line.trim().split("\t")
+          println("context words for.." + dbpediaURI + " will be deleted")
+          println("context words for.." + dbpediaURI + " will be set as given in input")
```
`"context words for.." + dbpediaURI + " will deleted and then set as given in input"`?

papaonlegs commented on src/main/scala/org/idio/dbpedia/spotlight/SpotlightModelReader.scala:
```diff
@@ -1,180 +1,225 @@
-                modelUpdater.loadContextWords()
-              }
+        sourceFile.getLines().foreach { line =>
+          val splittedLine = line.trim().split("\t")
```
splitLine*

papaonlegs commented on src/main/scala/org/idio/dbpedia/spotlight/SpotlightModelReader.scala:
```diff
@@ -1,180 +1,225 @@
+        spotlightModelReader.exportContextStore(pathToFile)
+      }
 
+      // outputs the properties for 40 Surface forms.
```
You can let the user specify how many forms :)

papaonlegs commented on src/main/scala/org/idio/dbpedia/spotlight/SpotlightModelReader.scala:
```diff
@@ -1,180 +1,225 @@
-    }else{
 
-        action match{
+      //checks existence of Dbpedia's Ids, SF, and links between SF's and Dbpedia's ids.
```
... and then removes them

papaonlegs commented on src/main/scala/org/idio/dbpedia/spotlight/SpotlightModelReader.scala:
```diff
@@ -1,180 +1,225 @@
+
+
+
+
```
4 new lines?!?!?!

papaonlegs replied:
Blasphemy!

papaonlegs commented on src/main/scala/org/idio/dbpedia/spotlight/stores/CustomCandidateMapStore.scala:
```diff
@@ -0,0 +1,201 @@
+    } catch {
+      case e: Exception => {
+
+        println("\tcreating candidate map array for " + surfaceFormID)
```
Is this more like setting size while the above is actually creating candidate map array?

papaonlegs commented on src/main/scala/org/idio/dbpedia/spotlight/stores/CustomCandidateMapStore.scala:
```diff
@@ -0,0 +1,201 @@
+  /**
+   *  Drops the 'i'th element of a list
+   */
+  def dropIndex(xs: Array[Int], n: Int): Array[Int] = {
```
This feels so functional. :+1:

papaonlegs commented on src/main/scala/org/idio/dbpedia/spotlight/stores/CustomContextStore.scala:
```diff
@@ -0,0 +1,83 @@
+
+    val indexOftokenId = this.contextStore.tokens(dbpediaResourceID).indexWhere { case (x) => x == tokenID }
+    //remove from the tokens array
+    this.contextStore.tokens(dbpediaResourceID) = this.contextStore.tokens(dbpediaResourceID).take(indexOftokenId) ++ this.contextStore.tokens(dbpediaResourceID).drop(indexOftokenId + 1)
```
dropIndex

papaonlegs commented on src/main/scala/org/idio/dbpedia/spotlight/stores/CustomContextStore.scala:
```diff
@@ -0,0 +1,83 @@
+    this.contextStore.tokens(dbpediaResourceID) = this.contextStore.tokens(dbpediaResourceID).take(indexOftokenId) ++ this.contextStore.tokens(dbpediaResourceID).drop(indexOftokenId + 1)
+
+    //remove from counts array
+    this.contextStore.counts(dbpediaResourceID) = this.contextStore.counts(dbpediaResourceID).take(indexOftokenId) ++ this.contextStore.counts(dbpediaResourceID).drop(indexOftokenId + 1)
```
dropIndex

papaonlegs commented on src/main/scala/org/idio/dbpedia/spotlight/stores/CustomDbpediaResourceStore.scala:
```diff
@@ -0,0 +1,108 @@
+  * */
+  private def addDbpediaURI(uri: String, support: Int, types: Array[String]) {
+    //URI i.e: Click-through_rate
+    //Types: ??
```
??

dav009 replied:
:[ we are ignoring `types` at the moment , I will replace the ??

papaonlegs commented on src/main/scala/org/idio/dbpedia/spotlight/stores/CustomDbpediaResourceStore.scala:
```diff
@@ -0,0 +1,108 @@
+  }
+
+  /*
+  * Adds a new Dbpedia Resource to teh store, and rebuilds the indexes.
```
teh

papaonlegs commented on src/main/scala/org/idio/dbpedia/spotlight/stores/CustomSurfaceFormStore.scala:
```diff
@@ -0,0 +1,198 @@
+    // look for existing surfaceForm
+    try {
+      var surfaceForm = this.sfStore.getSurfaceForm(surfaceText)
+      this.boostCountsIfNeeded(surfaceForm.id)
```
Why do you need to `boostCountsIfNeeded` here since the surfaceForm already exists and is most likely OK.

If this needs to stay I'd suggest making a note in the docbloc saying existing SFs may be boosted

dav009 replied:
It can exist but it might not have enough probability to show up when spotting text.

Assumming "most likely is OK" is a bad assumption. However `boostCountsIfNeeded` would only change the probability if it is below the min threshold. The responsibility of this function is to force something to be spottable.

this is already in the docs. In the diagnosis guide.

papaonlegs replied:
Cool

papaonlegs commented on src/main/scala/org/idio/dbpedia/spotlight/stores/CustomTokenResourceStore.scala:
```diff
@@ -0,0 +1,131 @@
+    this.tokenStore.tokenForId = this.tokenStore.tokenForId :+ token
+    this.tokenStore.counts = this.tokenStore.counts :+ 1
+
+    return this.tokenStore.tokenForId.size - 1
```
This returns the size of the `tokenForId` array while the prior return returns the `id`, will the added token always have id same as the size of the array?

dav009 replied:
yes, basically the id is the position in the array. 
If it is exists it returns the value, otherwise it puts it at the end of the array so its id is the last index.

papaonlegs replied:
Cool

papaonlegs commented on src/main/scala/org/idio/dbpedia/spotlight/stores/CustomTokenResourceStore.scala:
```diff
@@ -0,0 +1,131 @@
+  /*
+  * Returns a stemmed version of a Token.
+  * */
+  def stemToken(token: String): String = {
```
Why no function for `snowballStemToken`? Not that I know why it's used.

papaonlegs replied:
NVM me. The stemmer IS a `snowballStemmer`

papaonlegs commented on src/main/scala/org/idio/dbpedia/spotlight/stores/CustomTokenResourceStore.scala:
```diff
@@ -0,0 +1,131 @@
+    (contextWords, contextCounts).zipped foreach { (word, counts) =>
+      {
+        val stemmedWord: String = this.stemToken(word)
+        var currentCount = 0
```
You should define this outside the loop and just set to 0 here, should make a small difference. Plus didn't you say in the guide to avoid using `var` and instead `val`?

dav009 replied:
I think I can get rid of the `var` here now, I wrote this long ago.

basically all of this (conidtional + var + assigment) could be replaced for: 

```
val currentCount =  contextTokenMap.getOrElse(stemmedWord, 0)
```

:+1:

papaonlegs commented on src/main/scala/org/idio/dbpedia/spotlight/utils/CommandLineParser.scala:
```diff
@@ -0,0 +1,157 @@
+
+
+
+
```
5 new lines!?!?!

papaonlegs replied:
Abomination!

papaonlegs commented on src/main/scala/org/idio/dbpedia/spotlight/utils/ModelUpdateFromFile.scala:
```diff
@@ -1,122 +1,90 @@
+  def loadNewEntriesFromFile() {
+
+    val fileParser = new ModelFileParser(this.pathToFile)
+    println("Parsing INPUT-FILE")
```
Wanna make INPUT-FILE `this.pathToFile` ?

papaonlegs commented on src/main/scala/org/idio/dbpedia/spotlight/utils/ModelUpdateFromFile.scala:
```diff
@@ -1,122 +1,90 @@
 
-    println("adding all SF..."+ setOfSurfaceForms.size)
-    idioSpotlightModel.addListOfSurfaceForms(setOfSurfaceForms)
+    println("Finished parsing INPUT-FILE")
```
Same as above

papaonlegs commented on src/main/scala/org/idio/dbpedia/spotlight/utils/ModelUpdateFromFile.scala:
```diff
@@ -1,122 +1,90 @@
 
-        val (surfaceFormId, dbpediaResourceId) = idioSpotlightModel.addNew(surfaceForm,dbpediaURI, types, contextWordsArray, contextCounts )
+        // Updates the model connecting Sf-> Topic
+        // Topic -> Contxt Words
```
Context*

dav009 committed:
cleaning licenses/readme

dav009 committed:
cleaning comments and prints

dav009 committed:
better getcontext tokens

dav009 committed:
sharing code between unspottable and spottable commands

dav009 committed:
better array index dropping

dav009 committed:
better variable names

dav009 committed:
fixing comments

dav009 committed:
adding param to explore cmd

dav009 committed:
cleaning more comments

dav009 committed:
removing sensible information from readme

papaonlegs commented:
Looks like it's ready to be merged.
![](http://stream1.gifsoup.com/view6/20140519/5039131/principediego-a.gif)

dav009 commented:
YAY!!!!
